runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java

```diff
@@ -108,38 +115,57 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
 
   @Override
   public SparkPipelineResult run(Pipeline pipeline) {
-    // clear state of Aggregators, Metrics and Watermarks if exists.
+    SparkPipelineOptions sparkOptions = pipeline.getOptions().as(SparkPipelineOptions.class);
+    SparkPipelineResult result = null;
+    // clear state of Aggregators, Metrics and Watermarks.
     AggregatorsAccumulator.clear();
     SparkMetricsContainer.clear();
     GlobalWatermarkHolder.clear();
 
     TestPipelineOptions testPipelineOptions = pipeline.getOptions().as(TestPipelineOptions.class);
-    SparkPipelineResult result = delegate.run(pipeline);
-    result.waitUntilFinish();
-
-
-    // make sure the test pipeline finished successfully.
-    State resultState = result.getState();
-    assertThat(
-        String.format("Test pipeline result state was %s instead of %s", resultState, State.DONE),
-        resultState,
-        is(State.DONE));
-    assertThat(result, testPipelineOptions.getOnCreateMatcher());
-    assertThat(result, testPipelineOptions.getOnSuccessMatcher());
+    LOG.info("About to run test pipeline " + sparkOptions.getJobName());
 
     // if the pipeline was executed in streaming mode, validate aggregators.
     if (isForceStreaming) {
-      // validate assertion succeeded (at least once).
-      int success = result.getAggregatorValue(PAssert.SUCCESS_COUNTER, Integer.class);
-      assertThat(
-          String.format(
-              "Expected %d successful assertions, but found %d.",
-              expectedNumberOfAssertions, success),
-          success,
-          is(expectedNumberOfAssertions));
-      // validate assertion didn't fail.
-      int failure = result.getAggregatorValue(PAssert.FAILURE_COUNTER, Integer.class);
-      assertThat("Failure aggregator should be zero.", failure, is(0));
+      try {
+        result = delegate.run(pipeline);
+        long timeout = sparkOptions.getForcedTimeout();
+        result.waitUntilFinish(Duration.millis(timeout));
+        // validate assertion succeeded (at least once).
+        int successAssertions = result.getAggregatorValue(PAssert.SUCCESS_COUNTER, Integer.class);
+        assertThat(
+            String.format(
+                "Expected %d successful assertions, but found %d.",
+                expectedNumberOfAssertions, successAssertions),
+            successAssertions,
+            is(expectedNumberOfAssertions));
+        // validate assertion didn't fail.
+        int failedAssertions = result.getAggregatorValue(PAssert.FAILURE_COUNTER, Integer.class);
+        assertThat(
```

Comment:
This is "just in-case". The assertion errors should show in the log, and most of the time fail the test entirely. I did see in the past some weird behaviour where a failure occurs but the job keeps on going (logging the error and all) so this is just to be on the safe side.
So from my experience the error would log anyway, but this is a safety-net for silent failures.